# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Load BAM header
#'
#' This function loads the header of a BAM file into a List.  It will attempt to load the
#' list of sequences, program used and read groups as well as the version of the
#' BAM file, skipping over any information that is not present.
#'
#' @param filename The BAM file name
#' @export
#' @examples loadheader("~git/pbbam/tests/data/dataset/bam_mapping_1.bam")
loadheader <- function(filename) {
    .Call('pbbamr_loadheader', PACKAGE = 'pbbamr', filename)
}

#' Load PBI BAM index file
#'
#' This function loads a pbi index file into a dataframe.  Depending on the
#' number of attributes present, it will either load just the basic data, or optionally
#' the mapping and barcode data.
#'
#' The original BAM file can also be read to gather additional covariates such as the SNR, read quality
#' and number of passes, though this may take longer.
#'
#' @param filename The BAM file name (without .pbi)
#' @param loadSNR Should we load the four channel SNR data? (Default = FALSE)
#' @param loadNumPasses Should we load the number of passes data? (Default = FALSE)
#' @param loadRQ Shouold we load the read quality? (Default = FALSE)
#' @export
#' @examples loadpbi("~git/pbbam/tests/data/dataset/bam_mapping_1.bam")
loadpbi <- function(filename, loadSNR = FALSE, loadNumPasses = FALSE, loadRQ = FALSE) {
    .Call('pbbamr_loadpbi', PACKAGE = 'pbbamr', filename, loadSNR, loadNumPasses, loadRQ)
}

#' Load BAM alignments as a list of data frames.
#'
#' @param offsets The virtual file offsets to retrieve BAM records from (can be obtained from the index file based on loadpbi).
#' @param bamName The BAM file name to grab
#' @param indexedFastaName The name of the indexed fasta file this should come from.
#'
#' @return Returns a list of alignments as data frames.  If the IPD and Pulse Width are available, they will be columns in the returned data as well.
#' @export
loadDataAtOffsets <- function(offsets, bamName, indexedFastaName) {
    .Call('pbbamr_loadDataAtOffsets', PACKAGE = 'pbbamr', offsets, bamName, indexedFastaName)
}

#' Load BAM subreads as a list of data frames.
#'
#' @param offsets The virtual file offsets to retrieve BAM records from (can be obtained from the index file based on loadpbi).
#' @param bamName The BAM file name to grab
#'
#' @return Returns a list of subreads as a list of character strings.
#' @export
loadSubreadsAtOffsets <- function(offsets, bamName) {
    .Call('pbbamr_loadSubreadsAtOffsets', PACKAGE = 'pbbamr', offsets, bamName)
}

#' Load a section of the reference genome as a character string.
#'
#' @param id The name of the sequence
#' @param start The start of the sequence
#' @param end The end of the sequence
#' @param indexedFastaName The name of the indexed fasta file this should come from.
#'
#' @return Returns a character vector of the sequence
#' @export
loadReferenceWindow <- function(id, start, end, indexedFastaName) {
    .Call('pbbamr_loadReferenceWindow', PACKAGE = 'pbbamr', id, start, end, indexedFastaName)
}

#' Load BAM alignments as a list of list for the HMM model.
#'
#' @param offsets The virtual file offsets to retrieve BAM records from (can be obtained from the index file based on loadpbi).
#' @param bamName The BAM file name to grab
#' @param indexedFastaName The name of the indexed fasta file this should come from.
#' @param trimToLength How much should we subsample the alignments?
#'
#' @return Returns a list of phase2datasets as data frames.
#' @export
loadHMMfromBAM <- function(offsets, bamName, indexedFastaName, trimToLength = 140L) {
    .Call('pbbamr_loadHMMfromBAM', PACKAGE = 'pbbamr', offsets, bamName, indexedFastaName, trimToLength)
}

#' Load BAM alignment from a single ZMW as a list of list for the HMM model.
#'
#' @param offsets The virtual file offsets to retrieve BAM records from (can be obtained from the index file based on loadpbi).
#' @param bamName The BAM file name to grab
#' @param indexedFastaName The name of the indexed fasta file this should come from.
#' @param windowBreakSize We generate a new "window" every time 2 basepairs are matching in a particular gap.
#' @param minSize What is the minimum window size necessary for return, if the window at the end is less than this, we drop it. (NOT CURRENTLY USED).
#' @return Returns a list of phase2datasets as data frames.
#' @export
loadSingleZmwHMMfromBAM <- function(offsets, bamName, indexedFastaName, windowBreakSize = 140L, minSize = 50L) {
    .Call('pbbamr_loadSingleZmwHMMfromBAM', PACKAGE = 'pbbamr', offsets, bamName, indexedFastaName, windowBreakSize, minSize)
}

